### PR TITLE
[SPARK-44431][SQL] Fix behavior of null IN (empty list) in optimization rules

### PR DESCRIPTION
**Pull request by jchen5**
### What changes were proposed in this pull request?
`null IN (empty list)` incorrectly evaluates to null, when it should evaluate to false. (The reason it should be false is because a IN (b1, b2) is defined as a = b1 OR a = b2, and an empty IN list is treated as an empty OR which is false. This is specified by ANSI SQL.)

Many places in Spark execution (In, InSet, InSubquery) and optimization (OptimizeIn, NullPropagation) implemented this wrong behavior. This is a longstanding correctness issue which has existed since null support for IN expressions was first added to Spark.

This PR fixes the optimization rules OptimizeIn and NullPropagation, which followed the preexisting, incorrect execution behavior. The execution fixes will be in the next PR.

The behavior is under a flag, which will be available to revert to the legacy behavior if needed. This flag is set to disable the new behavior until all of the fix PRs are complete.

See [this doc](https://docs.google.com/document/d/1k8AY8oyT-GI04SnP7eXttPDnDj-Ek-c3luF2zL6DPNU/edit) for more information.

### Why are the changes needed?
Fix wrong SQL semantics

### Does this PR introduce _any_ user-facing change?
Not yet, but will fix wrong SQL semantics when enabled

### How was this patch tested?
Add unit tests and sql tests.

### CODE REVIEW

**Comment by jchen5** [comment]
@cloud-fan

File: sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-null-semantics.sql [hunk 1]
```diff
@@ -0,0 +1,57 @@
+create temp view v (c) as values (1), (null);
+create temp view v_empty (e) as select 1 where false;
+
+-- Note: tables and temp views hit different optimization/execution codepaths
+create table t(c int) using parquet;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-null-semantics.sql:5]: nit: if file format doesn't matter, probably json/csv is better as it's faster to create and read small tables.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-null-semantics.sql (hunk 1 above)
Comment (jchen5) [reply sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-null-semantics.sql:5]: 👍 will update

**Comment by jchen5** [comment]
@cloud-fan PR is updated

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**